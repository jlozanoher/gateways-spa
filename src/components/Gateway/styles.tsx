import styled from "styled-components";
import * as Antd from "antd";

export const Header = styled.div`
  display: flex;
  align-items: center;
  padding: 8px 16px;
`;

export const Grow = styled.div`
  flex-grow: 1;
`;

export const Modal = styled(Antd.Modal)`
  padding: 40px 24px;
`;
