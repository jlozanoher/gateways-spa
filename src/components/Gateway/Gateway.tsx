import { DeleteOutlined } from "@ant-design/icons";
import { Button, Typography } from "antd";
import React, { useContext } from "react";
import { useDrop } from "react-dnd";
import { PeripheralContext } from "../../App";
import { useGatewayDelete } from "../../contollers/gateway.controller";
import { usePeripheralUpdate } from "../../contollers/peripheral.controller";
import { dotsString } from "../../lib/utils";
import { GatewayModel } from "../../types/gateway.model";
import { ItemTypes } from "../../types/ItemsTypes";
import { PeripheralList } from "../Peripheral/PeripheralList";
import * as S from "../styles";
import { GatewayContext } from "./Gateways";

const { Title, Text } = Typography;

interface Props {
  gateway: GatewayModel;
}

export const Gateway = (props: Props) => {
  const { gateway } = props;

  const { fetchGateways = () => {} } = useContext(GatewayContext);
  const { peripherals, fetchPeripherals = () => {} } =
    useContext(PeripheralContext);

  const { handleUpdatePeripheral } = usePeripheralUpdate({
    onSuccess: () => {
      fetchPeripherals();
    },
  });

  const [collected, drop] = useDrop(
    () => ({
      accept: ItemTypes.PERIPHERAL,
      drop: (item: any, monitor: any) => {
        handleUpdatePeripheral({ ...item, gateway: gateway._id });
      },
      collect: (monitor) => {
        return monitor.getItem();
      },
    }),
    []
  );

  const { handleDeleteGateway } = useGatewayDelete({
    onSuccess: () => {
      fetchGateways();
      fetchPeripherals();
    },
  });

  const onDelete = () => {
    handleDeleteGateway(gateway._id);
  };

  const attachedPeripherals = peripherals?.filter(
    (p) => p.gateway === gateway._id
  );

  return (
    <div ref={drop}>
      <S.Card>
        <S.Box>
          <S.Box>
            <Button icon={<DeleteOutlined />} danger onClick={onDelete} />
          </S.Box>
          <S.Box style={{ flexDirection: "column" }}>
            <Title level={5}>{gateway.name}</Title>
            <Text type="secondary">{dotsString(gateway._id)}</Text>
            <div>{gateway.serialNumber}</div>
            <div>{gateway.ipv4Address}</div>
          </S.Box>
          <S.Box style={{ flexDirection: "column", flexGrow: 1 }}>
            <Title level={5}>{"Peripherals"}</Title>
            <PeripheralList peripherals={attachedPeripherals || []} />
          </S.Box>
        </S.Box>
      </S.Card>
    </div>
  );
};
