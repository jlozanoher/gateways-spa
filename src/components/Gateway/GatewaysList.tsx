import React from "react";
import * as S from "../styles";

import { GatewayModel } from "../../types/gateway.model";
import { Gateway } from "./Gateway";

interface Props {
  gateways: GatewayModel[];
}

export const GatewaysList = (props: Props) => {
  const { gateways } = props;

  return (
    <S.DivH>
      {gateways.map((gateway) => (
        <Gateway key={gateway._id} gateway={gateway} />
      ))}
    </S.DivH>
  );
};
