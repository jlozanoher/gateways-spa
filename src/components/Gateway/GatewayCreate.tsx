import { Button, Form, Input } from "antd";
import React, { useContext, useEffect } from "react";
import {
  useGatewayCreate,
  useGatewayUpdate,
} from "../../contollers/gateway.controller";
import { GatewayModel } from "../../types/gateway.model";
import { GatewayContext } from "./Gateways";
import * as S from "../styles";

const layout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 16 },
};
const tailLayout = {
  wrapperCol: { offset: 8, span: 16 },
};

interface Props {
  gateway?: GatewayModel; // For updating instead of creating
}

export const GatewayCreate = (props: Props) => {
  const { gateway } = props;

  const { fetchGateways = () => {} } = useContext(GatewayContext);

  const { handleCreateGateway } = useGatewayCreate({
    onSuccess: () => {
      fetchGateways();
    },
    onError: (err) => {
      if (err?.name?.code === 11000 && err?.name?.keyPattern?.serialNumber)
        form.setFields([
          {
            name: "serialNumber",
            errors: ["Duplicity"],
          },
        ]);
    },
  });
  const { handleUpdateGateway } = useGatewayUpdate({
    onSuccess: () => {
      fetchGateways();
    },
  });

  const [form] = Form.useForm();

  useEffect(() => {
    if (gateway) {
      form.setFieldsValue(gateway);
    }
  }, [gateway]);

  const onFinish = (values: any) => {
    if (gateway?._id) {
      handleUpdateGateway({ ...values, _id: gateway._id });
    } else {
      handleCreateGateway(values);
    }
  };

  const onClear = () => {
    form.resetFields();
  };

  return (
    <div>
      <Form {...layout} form={form} onFinish={onFinish}>
        <Form.Item name={"name"} label={"Name"} rules={[{ required: true }]}>
          <Input />
        </Form.Item>
        <Form.Item
          name={"serialNumber"}
          label={"Serial number"}
          rules={[{ required: true }]}
        >
          <Input />
        </Form.Item>
        <Form.Item
          name={"ipv4Address"}
          label={"IPv4"}
          rules={[
            { required: true },
            () => ({
              validator(rule, value) {
                if (
                  !/^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/.test(
                    value
                  )
                ) {
                  return Promise.reject("Invalid IPv4 address!");
                }
                return Promise.resolve();
              },
            }),
          ]}
        >
          <Input />
        </Form.Item>
        <Form.Item {...tailLayout}>
          <S.Box>
            <Button type="primary" htmlType="submit">
              {gateway ? "Update Gateway" : "Create Gateway"}
            </Button>
            <Button onClick={onClear}>Clear</Button>
          </S.Box>
        </Form.Item>
      </Form>
    </div>
  );
};
