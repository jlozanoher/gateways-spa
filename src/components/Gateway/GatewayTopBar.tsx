import { PlusOutlined, WarningOutlined } from "@ant-design/icons";
import { Button, Typography } from "antd";
import React, { useContext } from "react";
import { GatewayContext } from "./Gateways";
import * as S from "../styles";

const { Title, Text } = Typography;

export const GatewayTopBar = () => {
  const { setShowModal = () => {} } = useContext(GatewayContext);

  const onClick = () => {
    setShowModal(true);
  };

  return (
    <S.Header>
      <Title level={3}>Gateways</Title>
      <S.Grow>
        <Text type="warning">
          <WarningOutlined />
          Drag the peripherals into or out a Gateway
          <WarningOutlined />
        </Text>
      </S.Grow>
      <Button type="primary" icon={<PlusOutlined />} onClick={onClick}>
        Add Gateway
      </Button>
    </S.Header>
  );
};
