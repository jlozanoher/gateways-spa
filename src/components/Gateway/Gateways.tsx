import { Button } from "antd";
import axios from "axios";
import React, { Dispatch, SetStateAction, useEffect, useState } from "react";
import { ROUTES } from "../../lib/routes";
import { GatewayModel } from "../../types/gateway.model";
import * as S from "../styles";
import { GatewayCreate } from "./GatewayCreate";
import { GatewaysList } from "./GatewaysList";
import { GatewayTopBar } from "./GatewayTopBar";

export interface IGatewayContext {
  fetchGateways?: () => void;
  setShowModal?: Dispatch<SetStateAction<boolean>>;
  showModal?: boolean;
  gatewayForEditon?: GatewayModel;
  setGatewayForEdition?: Dispatch<SetStateAction<GatewayModel | undefined>>;
}

export const GatewayContext = React.createContext<IGatewayContext>({});

export const Gateways = () => {
  const [gateways, setGateways] = useState<GatewayModel[]>([]);
  const [gatewayForEditon, setGatewayForEdition] = useState<
    GatewayModel | undefined
  >(undefined);
  const [showModal, setShowModal] = useState(false);

  const fetchGateways = () => {
    axios
      .get(`${process.env.REACT_APP_API_URL}${ROUTES.gateways}`)
      .then((res) => {
        setGateways(res.data);
      });
  };

  useEffect(() => {
    fetchGateways();
  }, []);

  const handleModalClose = () => {
    setShowModal(false);
    setGatewayForEdition(undefined);
  };

  return (
    <GatewayContext.Provider
      value={{
        fetchGateways,
        setShowModal,
        showModal,
        gatewayForEditon,
        setGatewayForEdition,
      }}
    >
      <GatewayTopBar />
      <GatewaysList gateways={gateways} />
      <S.Modal
        title={gatewayForEditon ? "Update Gateway" : "Create Gateway"}
        maskClosable
        visible={showModal}
        onCancel={handleModalClose}
        footer={[
          <Button type="primary" onClick={handleModalClose}>
            Cancel
          </Button>,
        ]}
      >
        <GatewayCreate gateway={gatewayForEditon} />
      </S.Modal>
    </GatewayContext.Provider>
  );
};
