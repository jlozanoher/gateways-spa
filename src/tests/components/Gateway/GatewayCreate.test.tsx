import React from "react";
import renderer from "react-test-renderer";
import { GatewayCreate } from "../../../components/Gateway/GatewayCreate";

global.matchMedia =
  global.matchMedia ||
  function () {
    return {
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  };

describe("GatewayCreate should match snaps", () => {
  it("GatewayCreate", () => {
    const tree = renderer.create(<GatewayCreate />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("GatewayCreate in edit mode", () => {
    const tree = renderer
      .create(
        <GatewayCreate
          gateway={{
            _id: "1231da",
            ipv4Address: "1.1.1.1",
            name: "Gate 1",
            serialNumber: "123",
          }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
