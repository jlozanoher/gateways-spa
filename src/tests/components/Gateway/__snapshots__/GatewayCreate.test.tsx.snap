// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GatewayCreate should match snaps GatewayCreate 1`] = `
<div>
  <form
    className="ant-form ant-form-horizontal"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="ant-row ant-form-item"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-6 ant-form-item-label"
        style={Object {}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="name"
          title="Name"
        >
          Name
        </label>
      </div>
      <div
        className="ant-col ant-col-16 ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <input
              className="ant-input"
              id="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-6 ant-form-item-label"
        style={Object {}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="serialNumber"
          title="Serial number"
        >
          Serial number
        </label>
      </div>
      <div
        className="ant-col ant-col-16 ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <input
              className="ant-input"
              id="serialNumber"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-6 ant-form-item-label"
        style={Object {}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="ipv4Address"
          title="IPv4"
        >
          IPv4
        </label>
      </div>
      <div
        className="ant-col ant-col-16 ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <input
              className="ant-input"
              id="ipv4Address"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-16 ant-col-offset-8 ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <div
              className="sc-hKwDye dbURyo"
            >
              <button
                className="ant-btn ant-btn-primary"
                onClick={[Function]}
                type="submit"
              >
                <span>
                  Create Gateway
                </span>
              </button>
              <button
                className="ant-btn"
                onClick={[Function]}
                type="button"
              >
                <span>
                  Clear
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`GatewayCreate should match snaps GatewayCreate in edit mode 1`] = `
<div>
  <form
    className="ant-form ant-form-horizontal"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="ant-row ant-form-item ant-form-item-has-success"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-6 ant-form-item-label"
        style={Object {}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="name"
          title="Name"
        >
          Name
        </label>
      </div>
      <div
        className="ant-col ant-col-16 ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <input
              className="ant-input"
              id="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="text"
              value="Gate 1"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item ant-form-item-has-success"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-6 ant-form-item-label"
        style={Object {}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="serialNumber"
          title="Serial number"
        >
          Serial number
        </label>
      </div>
      <div
        className="ant-col ant-col-16 ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <input
              className="ant-input"
              id="serialNumber"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="text"
              value="123"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item ant-form-item-has-success"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-6 ant-form-item-label"
        style={Object {}}
      >
        <label
          className="ant-form-item-required"
          htmlFor="ipv4Address"
          title="IPv4"
        >
          IPv4
        </label>
      </div>
      <div
        className="ant-col ant-col-16 ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <input
              className="ant-input"
              id="ipv4Address"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="text"
              value="1.1.1.1"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-row ant-form-item"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-16 ant-col-offset-8 ant-form-item-control"
        style={Object {}}
      >
        <div
          className="ant-form-item-control-input"
        >
          <div
            className="ant-form-item-control-input-content"
          >
            <div
              className="sc-hKwDye dbURyo"
            >
              <button
                className="ant-btn ant-btn-primary"
                onClick={[Function]}
                type="submit"
              >
                <span>
                  Update Gateway
                </span>
              </button>
              <button
                className="ant-btn"
                onClick={[Function]}
                type="button"
              >
                <span>
                  Clear
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
